File: May-LeetCoding-Challange-2020/Week4/1035-uncrossed-lines.js
/**
 * @param {number[]} A
 * @param {number[]} B
 * @return {number}
 * Runtime: 84 ms, faster than 25.00% of JavaScript online submissions for Uncrossed Lines.
 * Memory Usage: 41.5 MB, less than 66.67% of JavaScript online submissions for Uncrossed Lines.
 */
var maxUncrossedLines = function(A, B) {
    const aLength = A.length;
    const bLength = B.length;
    const dp = [...Array(aLength+1)].map(x => Array(bLength+1).fill(0));

    for(let aIndex = 1; aIndex <= aLength; aIndex++) {
        for(let bIndex = 1; bIndex <= bLength; bIndex++) {
            if (A[aIndex-1] === B[bIndex-1]) {
                dp[aIndex][bIndex] = dp[aIndex - 1][bIndex - 1] + 1;
            } else {
                dp[aIndex][bIndex] = Math.max(dp[aIndex][bIndex - 1], dp[aIndex - 1][bIndex]);
            }
        }
    }

    return dp[aLength][bLength];
};

console.log(maxUncrossedLines([1, 4, 2],[1, 2, 4]));
console.log(maxUncrossedLines([3, 3],[1, 3, 1, 2, 1]));
console.log(maxUncrossedLines([1,3,7,1,7,5],[1,9,2,5,1]));

File: May-LeetCoding-Challange-2020/Week4/1035-uncrossed-lines.py
class Solution:
    def maxUncrossedLines(self, A: List[int], B: List[int]) -> int:
        N = len(A)
        M = len(B)
        
        dp = [[0 for n in range(N+1)] for m in range(M+1)]
        
        for j in range(1,M+1):
            for i in range(1, N+1):
                if A[i-1] == B[j-1]:
                    dp[j][i] = 1 + dp[j-1][i-1]
                else:
                    dp[j][i] = max(dp[j][i-1], dp[j-1][i])
        return dp[j][i]

        # Time Complexity: O(n*m), where n, m are the size of A and B
        # Space Complexity: O(n+1 * m+1) => O(n*m)
        # Submit Runtime: 244 ms, 58.08% faster
        # Submit Memory: 13.9 MB, 46.81% less
                